pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -0,0 +1,259 @@
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        uint256 numTokens = tokens.length;
+
+        normalizedWeights = new uint256[](numTokens);
+
+        uint256 pctProgress = _calculateWeightChangeProgress();
+
+        for (uint256 i = 0; i < numTokens; i++) {
+            bytes32 tokenData = _poolState[tokens[i]];
+
+            normalizedWeights[i] = _interpolateWeight(tokenData, pctProgress);
+        }
+    }
+
+    function _getNormalizedWeightsAndMaxWeightIndex()
+        internal
+        view
+        override
+        returns (uint256[] memory normalizedWeights, uint256 maxWeightTokenIndex)
+    {
+        normalizedWeights = _getNormalizedWeights();
+
+        maxWeightTokenIndex = 0;
+        uint256 maxNormalizedWeight = normalizedWeights[0];
+
+        for (uint256 i = 1; i < normalizedWeights.length; i++) {
+            if (normalizedWeights[i] > maxNormalizedWeight) {
+                maxWeightTokenIndex = i;
+                maxNormalizedWeight = normalizedWeights[i];
+            }
+        }
+    }
+
+    /**
+     * @dev When calling updateWeightsGradually again during an update, reset the start weights to the current weights,
+     * if necessary. Time travel elements commented out.
+     */
+    function _startGradualWeightChange(
+        uint256, // startTime,
+        uint256, // endTime,
+        uint256[] memory startWeights,
+        uint256[] memory endWeights,
+        IERC20[] memory tokens
+    ) internal virtual {
+        //bytes32 newTimestamps = _gradualUpdateTimestamps;
+        uint256 normalizedSum = 0;
+        bytes32 tokenState;
+
+        for (uint256 i = 0; i < endWeights.length; i++) {
+            uint256 endWeight = endWeights[i];
+            _require(endWeight >= _MIN_WEIGHT, Errors.MIN_WEIGHT);
+
+            IERC20 token = tokens[i];
+
+            // Tokens with more than 18 decimals are not supported
+            // Scaling calculations must be exact/lossless
+            // Store decimal difference instead of actual scaling factor
+            _poolState[token] = tokenState
+                .insertUint64(startWeights[i].compress64(), _START_WEIGHT_OFFSET)
+                .insertUint32(endWeight.compress32(), _END_WEIGHT_OFFSET)
+                .insertUint5(uint256(18).sub(ERC20(address(token)).decimals()), _DECIMAL_DIFF_OFFSET);
+
+            normalizedSum = normalizedSum.add(endWeight);
+        }
+        // Ensure that the normalized weights sum to ONE
+        _require(normalizedSum == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
+
+        //_gradualUpdateTimestamps = newTimestamps
+        //    .insertUint32(startTime, _START_TIME_OFFSET)
+        //    .insertUint32(endTime, _END_TIME_OFFSET);
+
+        //emit GradualWeightUpdateScheduled(startTime, endTime, startWeights, endWeights);
+    }
+
+    function _computeScalingFactor(bytes32 tokenState) private view returns (uint256) {
```

Comment:
Let's chose a different name: `_computeScalingFactor` already exists (for an argument of type `IERC20`) and does something completely different.

Reply:
computeScalingMultiplier?

Reply:
Well, maybe _readScalingFactor - it does the exact same thing as BasePool, except it reads from the mapping rather than calling the token.decimals()